lib/active_model/serializer/adapter/json_sideload.rb

```diff
@@ -0,0 +1,206 @@
+module ActiveModel
+  class Serializer
+    class Adapter
+      class JsonSideload < Json
+
+        # the list of what has already been serialized will be kept here to
+        # help avoid infinite recursion
+        #
+        # this should be a list of association_name to a list of objects
+        attr_accessor :serialized
+
+        # When we are sideloading associations, we can more easily track what
+        # has been serialized, so that we avoid infinite
+        # recursion / serialization.
+        def initialize(*args)
+          @serialized = {}
+
+          super(*args)
+        end
+
+        def serializable_hash(options = nil)
+          options ||= {}
+
+          # begin recursive serialization
+          # TODO: this will evaluate ALL specified relationships
+          # throughout the entire tree.
+          #
+          # Do we want a way to limit this?
+          serialize_hash(options)
```

Comment:
yes!. max depth of 3 or 4 should be good, imho. 

Related issues
- https://github.com/rails-api/active_model_serializers/pull/1073/files#r38325279 (definitely see this!)
- https://github.com/rails-api/active_model_serializers/issues/211#issuecomment-32900106
- https://github.com/rails-api/active_model_serializers/issues/540
- https://github.com/rails-api/active_model_serializers/pull/722
- https://github.com/rails-api/active_model_serializers/pull/962
- https://github.com/rails-api/active_model_serializers/commit/2e791b6d5b0d40ab4ad427ba0e02f84b78bb8345

I know rubygems has a good recursive resolution lib ([molinillo](https://github.com/rubygems/rubygems/blob/e3088dd686f0fa4f935ac34298eea9e4f563d13b/lib/rubygems/resolver/molinillo/lib/molinillo/resolver.rb))